### ISSUE TITLE
Poor performance on pretrained models

### ISSUE THREAD

**Issue by OliverT1** [KatarinaYuan/StructTokenBench#7]
Hi,
Thank you for your work here. I'm having difficulties using the pretrained models. I have downloaded the models from the Google Drive using the provided instructions, but get poor performance on any of the downstream tasks. E.g. here is the codebook utilisation for casp14. Are the uploaded models definitely the fully trained ones, or am I loading them wrong? For the other benchmarked models (e.g. ESM3 and foldseek) I get the expcted results.

Output form AminoAseed on CASP14 codebook utilisation:

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃      Validate metric       ┃        DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│             UR             │          0.249721          │
│          entropy           │          0.008936          │
│         perplexity         │          0.210583          │
│   test_compression_ratio   │    0.038902055472135544    │
│        test_entropy        │     4.575283527374268      │
│  test_entropy_normalized   │    0.008936100639402866    │
│         test_loss          │            0.0             │
│      test_perplexity       │     107.81841278076172     │
│ test_perplexity_normalized │    0.21058283746242523     │
│       test_use_ratio       │     0.2497209906578064     │

Bash script I used :
```
# using AminoAseed
ckpt_name="AminoAseed"
path="$DIR/struct_token_bench_release_ckpt/codebook_512x1024-1e+19-linear-fixed-last.ckpt/checkpoint/mp_rank_00_model_states.pt"
quantizer_use_linear_project=true


# general extra arguments besides $SHARED_ARGS
tokenizer=WrappedOurPretrainedTokenizer
tokenizername=ourpretrained_${ckpt_name}
d_model=1024
lr=0.001
quantizer_codebook_size=512

EXTRA_MODEL_ARGS="tokenizer_pretrained_ckpt_path=$path tokenizer_ckpt_name=${ckpt_name} quantizer_codebook_size=$quantizer_codebook_size quantizer_codebook_embed_size=$d_model model_encoder_dout=$d_model quantizer_use_linear_project=$quantizer_use_linear_project"



target_field=null task_goal="codebook_utilization" experiment_prefix="${task_goal}_casp14"
EXTRA_TASK_ARGS="test_only=true model.task_goal=${task_goal} experiment_name=${experiment_prefix}_${tokenizername} optimization.micro_batch_size=1"

SHARED_ARGS="tokenizer=$tokenizer model.d_model=$d_model trainer.devices=[0] optimization.optimizer.lr=$lr data.target_field=$target_field experiment_name=${experiment_prefix}_${tokenizername}_lr${lr} run_name=tryout_test default_data_dir=$DIR/struct_token_bench_release_data/ data.pdb_data_dir=$DIR/pdb_data/mmcif_files/ ${EXTRA_TASK_ARGS} ${EXTRA_MODEL_ARGS} trainer.default_root_dir=$DIR/struct_token_bench_logs/"
CUDA_VISIBLE_DEVICES=0 python ./src/script/run_supervised_task.py --config-name=casp14.yaml  $SHARED_ARGS
```

**Comment by KatarinaYuan** [comment]
Hi, can you try to output the codebook from your end?

**Comment by KatarinaYuan** [comment]
Hi,
   Thanks for your email and detecting this. In fact, there are several mis-written bugs due to refactoring for public release, which causes the following two inconsistency.

(1) For the old code, AminoAseed for "Sensitivity" task cannot match the paper. This is because that the codebook for AminoAseed should be `self.codebook.weight`, instead of `self.model.quantizer.linear_proj(self.model.quantizer.codebook.weight)`.  

(2) For the old code, the UR "unmatch" issue should happen for every baseline and AminoAseed. This is because that the `batch_size` should be 8 instead of 1 to reproduce the paper results. If you test ESM3 with batch_size=1, it leads to UR=0.0491. Changing batch_size to 8, the UR for ESM3 becomes 0.2760. That bing said, changing the batch_size from 1 to 8, would lead to the right UR for AminoAseed. 
Also, the idea is intuitive. More protein samples considered, the higher chance for one token index gets hit. This only leads to the change of the magnitude of UR, but the ordering for each method on this metric should be the same.

Now, code has been updated. Generally, thanks for interacting with our codebase and detecting the inconsistency. 

Best,
Xinyu

**Comment by david-heurtel-depeiges** [comment]
Hi, allow me to re-open potentially this issue. 

I believe all codebook utilization metrics should be aggregated over the whole test dataset (whether it is CASP or CAMEO). 

As previously mentionned, with the current implementation, the codebook utilization metrics are dependent on the batch size. I believe this is biasing larger models (for which a large batch size might not be possible) and model with large codebooks. UR is bounded by tokens_in_batch / codebook_size.  For example, with a batch size of 8 and assuming on average 200 residues in a protein, for a 4096-code codebook, the UR is capped at 0.4. Perplexity has the same issue via Jensen's inequality: mean(exp(H_batch)) < exp(H_global).

Beyond large models and large codebooks, this aggregation method is also penalizing models with correlated tokens in the sequence. I don't have a particular opinion on whether correlated structure tokens within a sequence are a good or bad thing (I would think it's reasonnable for all residues within a single alpha helix to be correlated). 

Suggested fix: concatenate the per-batch token tensors in _valid_or_test_epoch_end, call torch.bincount(..., minlength=V) once on the full stream, and compute use_ratio / entropy / perplexity from the global counts.

Thank you in advance for your anwer!
